src/main/python/pgshovel/replication/streams/kafka.py

```diff
@@ -40,12 +54,44 @@ def consume(self, state):
             consumer.seek(delta, 1)
             assert consumer.offsets[0] == offset
 
-        for offset, message in consumer:
-            decoded = self.codec.decode(message.value)
-            state = validate_state(state, offset, decoded)
+        for offset, message in self.__primed(consumer, state):
+            state = validate_state(state, offset, message)
             # XXX: This is necessary because of a bug in protocol buffer oneof.
             state = type(state).FromString(state.SerializeToString())
-            yield state, offset, decoded
+            yield state, offset, message
+
+    def __primed(self, consumer, state, attempts=5):
+        decoded = imap(
+            lambda (offset, msg): tuple((offset, self.codec.decode(msg.value))),
+            consumer
+        )
+
+        offset, message = next(decoded)
+
+        try:
+            validate_state(state, offset, message)
+        except InvalidEventError as e:
+            # See if we can drop the first few messages in an attempt to find
+            # the type we're looking for
+            while attempts > 0:
```

Comment:
> It seems like this would be very rare because the handoff from loader to stream should always occur on a transaction boundary. 

Given how Kafka topics are maintained I don't think it would be super-rare, but rare-ish yeah.  If you do run into this problem (and if this fix didn't exist) you would have to write sore application code to work around the stream state or munge the stream state manually - neither great.